### PR TITLE
`derive(Debug)` for `Expr`

### LINKED ISSUE #6677: `Debug` impl `Expr` is lossy
### Describe the bug

The `Debug` impl for `Expr` is a human-readable "nice" output:

https://github.com/apache/arrow-datafusion/blob/6872535358c2c6b70350484e6cc7c058c7a806ae/datafusion/expr/src/expr.rs#L872-L883

However it should be as detailed as possible. E.g. there's no way to tell from that output if `f(x)` is a function call or a column named `f(x)` which some of the optimizer passes do. This makes leads to an extremely frustrating debugging experience, esp. when some pass or step complains that a certain column wasn't found.

### To Reproduce

_No response_

### Expected behavior

Just auto-derive `Debug`.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by parkma99**
# Which issue does this PR close?

Closes #6677 

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by crepererum** [comment]
So I had a quick look at the outputs: I think the changes to the sqllogictests (e.g. `on=[(Column { name: "t1_id", index: 0 }, Column { name: "t2_id", index: 0 })]` => `on=[(t1_id@0, t2_id@0)]`) are desired. The new output is way easier to read and matches the formatting that we use for other plan formatting. Please update the tests accordingly.

**Comment by parkma99** [comment]
> So I had a quick look at the outputs: I think the changes to the sqllogictests (e.g. `on=[(Column { name: "t1_id", index: 0 }, Column { name: "t2_id", index: 0 })]` => `on=[(t1_id@0, t2_id@0)]`) are desired. The new output is way easier to read and matches the formatting that we use for other plan formatting. Please update the tests accordingly.

Ok

> [SQL] explain select * from hashjoin_datatype_table_t1 t1 right join hashjoin_datatype_table_t2 t2 on t1.c3 = t2.c3
[Diff] (-expected|+actual)
    logical_plan
    Right Join: CAST(t1.c3 AS Decimal128(10, 2)) = t2.c3
    --SubqueryAlias: t1
    ----TableScan: hashjoin_datatype_table_t1 projection=[c1, c2, c3, c4]
    --SubqueryAlias: t2
    ----TableScan: hashjoin_datatype_table_t2 projection=[c1, c2, c3, c4]
    physical_plan
    ProjectionExec: expr=[c1@0 as c1, c2@1 as c2, c3@2 as c3, c4@3 as c4, c1@5 as c1, c2@6 as c2, c3@7 as c3, c4@8 as c4]
-   --SortMergeJoin: join_type=Right, on=[(Column { name: "CAST(t1.c3 AS Decimal128(10, 2))", index: 4 }, Column { name: "c3", index: 2 })]
-   ----SortExec: expr=[CAST(t1.c3 AS Decimal128(10, 2))@4 ASC]
+   --SortMergeJoin: join_type=Right, on=[(Cast(Cast { expr: Column(Column { relation: Some(Bare { table: "t1" }), name: "c3" }), data_type: Decimal128(10, 2) })@4, c3@2)]
+   ----SortExec: expr=[Cast(Cast { expr: Column(Column { relation: Some(Bare { table: "t1" }), name: "c3" }), data_type: Decimal128(10, 2) })@4 ASC]
    ------CoalesceBatchesExec: target_batch_size=4096
-   --------RepartitionExec: partitioning=Hash([Column { name: "CAST(t1.c3 AS Decimal128(10, 2))", index: 4 }], 2), input_partitions=2
-   ----------ProjectionExec: expr=[c1@0 as c1, c2@1 as c2, c3@2 as c3, c4@3 as c4, CAST(c3@2 AS Decimal128(10, 2)) as CAST(t1.c3 AS Decimal128(10, 2))]
+   --------RepartitionExec: partitioning=Hash([Cast(Cast { expr: Column(Column { relation: Some(Bare { table: "t1" }), name: "c3" }), data_type: Decimal128(10, 2) })@4], 2), input_partitions=2
+   ----------ProjectionExec: expr=[c1@0 as c1, c2@1 as c2, c3@2 as c3, c4@3 as c4, CAST(c3@2 AS Decimal128(10, 2)) as Cast(Cast { expr: Column(Column { relation: Some(Bare { table: "t1" }), name: "c3" }), data_type: Decimal128(10, 2) })]
    ------------RepartitionExec: partitioning=RoundRobinBatch(2), input_partitions=1
    --------------MemoryExec: partitions=1, partition_sizes=[1]
    ----SortExec: expr=[c3@2 ASC]
    ------CoalesceBatchesExec: target_batch_size=4096
    --------RepartitionExec: partitioning=Hash([c3@2], 2), input_partitions=2
    ----------RepartitionExec: partitioning=RoundRobinBatch(2), input_partitions=1
    ------------MemoryExec: partitions=1, partition_sizes=[1]
at tests/sqllogictests/test_files/joins.slt:2672

Case 2 I find the reason. I will update right now.

File: datafusion/optimizer/src/analyzer/type_coercion.rs [hunk 1]
```diff
@@ -1040,7 +1040,7 @@ mod test {
         let empty = empty_with_type(DataType::Int64);
         let plan = LogicalPlan::Projection(Projection::try_new(vec![expr], empty)?);
         let expected =
-            "Projection: a IN ([CAST(Int32(1) AS Int64), CAST(Int8(4) AS Int64), Int64(8)]) AS a IN (Map { iter: Iter([Int32(1), Int8(4), Int64(8)]) })\
+            "Projection: a IN ([CAST(Int32(1) AS Int64), CAST(Int8(4) AS Int64), Int64(8)]) AS a IN (Map { iter: Iter([Literal(Int32(1)), Literal(Int8(4)), Literal(Int64(8))]) })\
              \n  EmptyRelation";
```
Comment (parkma99) [inline datafusion/optimizer/src/analyzer/type_coercion.rs:1044]: I think this test case should not change, but current I failed to past this test case.
When I change the code [0ef94](https://github.com/apache/arrow-datafusion/blob/0ef9478e6675308529d5420e94d70b0f3a32dc98/datafusion/expr/src/expr.rs#L1110) to use `Display` I got failed by `Projection: a IN ([CAST(Int32(1) AS Int64), CAST(Int8(4) AS Int64), Int64(8)])` missing `Alias`

Comment (crepererum) [reply datafusion/optimizer/src/analyzer/type_coercion.rs:1044]: Which alias?

File: datafusion/expr/src/expr.rs [hunk 2]
```diff
@@ -869,53 +869,56 @@ impl Expr {
     }
 }
 
-/// Format expressions for display as part of a logical plan. In many cases, this will produce
-/// similar output to `Expr.name()` except that column names will be prefixed with '#'.
-impl fmt::Display for Expr {
-    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
-        write!(f, "{self:?}")
-    }
+#[macro_export]
```
Comment (alamb) [inline datafusion/expr/src/expr.rs:872]: Given how common this functionality is I wonder if we could move this macro (or maybe make a function as I don't see any benefit in making it a macro)  in `datafusion_util` and use it more around the codebase.

Just a thought

**Review by alamb — COMMENTED** [review]
Thank you @parkma99  -- this is looking great and is very appreciated.

I can try and help with the formatting tomorrow if you are still having problems

File: datafusion/optimizer/src/analyzer/type_coercion.rs (hunk 1 above)
Comment (parkma99) [reply datafusion/optimizer/src/analyzer/type_coercion.rs:1044]: ` AS a IN (Map { iter: Iter([Int32(1), Int8(4), Int64(8)]) })\` is missing

File: datafusion/expr/src/expr.rs (hunk 2 above)
Comment (parkma99) [reply datafusion/expr/src/expr.rs:901]: I think making a function and moving to `datafusin_util` is a good option.

**Comment by alamb** [comment]
> When I run command cargo test in Terminal, it got error with message :
> `message: \"Too many open files\"`


That looks like you either have to increase your ulimit to increase the number of open files, or perhaps you can [reduce the parallelism of the tests](https://doc.rust-lang.org/book/ch11-02-running-tests.html#running-tests-in-parallel-or-consecutively): 

```shell
$ cargo test -- --test-threads=1
```

File: datafusion/core/tests/sqllogictests/test_files/tpch/q22.slt.part [hunk 3]
```diff
@@ -79,21 +79,21 @@ SortPreservingMergeExec: [cntrycode@0 ASC NULLS LAST]
...
 ------AggregateExec: mode=FinalPartitioned, gby=[cntrycode@0 as cntrycode], aggr=[COUNT(UInt8(1)), SUM(custsale.c_acctbal)]
 --------CoalesceBatchesExec: target_batch_size=8192
-----------RepartitionExec: partitioning=Hash([Column { name: "cntrycode", index: 0 }], 4), input_partitions=4
+----------RepartitionExec: partitioning=Hash([cntrycode@0], 4), input_partitions=4
 ------------AggregateExec: mode=Partial, gby=[cntrycode@0 as cntrycode], aggr=[COUNT(UInt8(1)), SUM(custsale.c_acctbal)]
 --------------ProjectionExec: expr=[substr(c_phone@0, 1, 2) as cntrycode, c_acctbal@1 as c_acctbal]
-----------------NestedLoopJoinExec: join_type=Inner, filter=BinaryExpr { left: CastExpr { expr: Column { name: "c_acctbal", index: 0 }, cast_type: Decimal128(19, 6), cast_options: CastOptions { safe: false, format_options: FormatOptions { safe: true, null: "", date_format: None, datetime_format: None, timestamp_format: None, timestamp_tz_format: None, time_format: None } } }, op: Gt, right: Column { name: "AVG(customer.c_acctbal)", index: 1 } }
+----------------NestedLoopJoinExec: join_type=Inner, filter=CAST(c_acctbal@0 AS Decimal128(19, 6)) > AVG(customer.c_acctbal)@1
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/tpch/q22.slt.part:85]: this is really much nicer 👍

File: datafusion/expr/src/expr.rs (hunk 2 above)
Comment (alamb) [reply datafusion/expr/src/expr.rs:901]: I can give it a try maybe after this PR is merged

**Review by alamb — APPROVED** [review]
This looks awesome @parkma99  -- thank you so much for making it happen

It is a big step forward for DataFusion usability I think

**Pull request merged**